### ISSUE TITLE
Distant Horizons

### ISSUE THREAD

**Issue by Madis0** [Madis0/fabulously-optimized#240]
### CurseForge link

https://www.curseforge.com/minecraft/mc-mods/distant-horizons

### CurseForge Mod Distribution

Allowed

### Modrinth link

https://modrinth.com/mod/distanthorizons

### Source/other link

https://gitlab.com/jeseibel/minecraft-lod-mod

### Mod file size

24.44 MiB

### What it does

Renders long distances with low detail, meaning anyone can have a long rendering distance

### Why should it be in the modpack

Actually exists for 1.18.1 Fabric unlike #86 and unlike #46 and #184 it is usable for low-end computers too

### Why shouldn't it be in the modpack

Alpha, untested by me, clainingly incompatible with (some) shaders, not sure if it works in multiplayer and whether it can cache chunks as well

### Categories

Performance optimization, Graphics optimization

### Additional details

https://youtu.be/PGmCPejswig

**Comment by TiboOpGithub** [comment]
I've tested it and it's a bit buggy and shader support is also lacking at the moment. When there is shader support it certainly seems like a good addition to FO! (8 render distane)
![2022-02-17_20 27 01](https://user-images.githubusercontent.com/95926002/154556073-a892ba90-e8f0-4da5-85ed-336fc6902f07.png)
![2022-02-17_20 27 02](https://user-images.githubusercontent.com/95926002/154556104-90f372ed-032b-4545-83ed-6da73e4e8665.png)

![2022-02-17_20 17 49](https://user-images.githubusercontent.com/95926002/154555998-618730d6-cd59-494a-8d1c-f47ed539ba73.png)

![Uploading 2022-02-17_20.20.46.png…
![2022-02-17_20 18 30](https://user-images.githubusercontent.com/95926002/154555991-95a6472e-cc8d-4723-8a3b-ddeae603bd5a.png)

![2022-02-17_20 20 44](https://user-images.githubusercontent.com/95926002/154555985-dfc400e2-c3c7-4b46-9b10-64c02c94ace8.png)

![2022-02-17_20 21 48](https://user-images.githubusercontent.com/95926002/154555905-0f97078e-696f-48a9-9829-110ab24d415c.png)
![2022-02-17_20 21 42](https://user-images.githubusercontent.com/95926002/154555919-1516f72b-acf4-402d-b4c5-8089b93716f9.png)
![2022-02-17_20 21 18](https://user-images.githubusercontent.com/95926002/154555965-8dad9656-e829-44e6-8e29-54d8ff073b44.png)
![2022-02-17_20 20 47](https://user-images.githubusercontent.com/95926002/154555971-939c1a94-9f22-42a1-9f91-06ba20b97209.png)

**Comment by Jackcool476** [comment]
After testing it in single-player worlds, it seems great. While it has no shader support as of yet and has some holes in the world during the initial chunk loading after joining a world, there is an extensive configuration that could probably fix the latter.

However, I think that its addition to FO should be postponed because of a fatal flaw:
![2022-02-18_15 50 42](https://user-images.githubusercontent.com/60570270/154716930-181d8132-d104-400a-8ecc-f87d46883b31.png)
This screenshot was taken in the Hypixel lobby after joining and leaving Skyblock. I have no idea how the mod works but I'd presume this issue has something to do with the way that you join and leave different worlds inside a single server. But the mod does seem to be great otherwise and if there was (or is) some way to stop its functionality in servers via its config then it would make for a great addition to FO right now.

**Comment by Probattler** [comment]
With my test, I loaded the hypixel duels lobby entirely, with high settings, low end pc, I was getting around 14-20 fps with random stutters, freezes and sometimes holes in the terrain as stated in previous comments, having worse performance than #46 and #184 in the same settings.

**Comment by osfanbuff63** [comment]
I will say that [Multiverse](https://github.com/Multiverse/Multiverse-Core) servers make this mod be very buggy, as it will show other worlds. Haven't tested it myself, or with nether/end though.

**Comment by JaredBr0wn** [comment]
It looks like they added support for servers with Multiverse in the [latest update](https://youtu.be/_04BZ8W2bDM?t=64).

**Comment by FixMagNum** [comment]
is it possible to run this mod on opengl 3.1 version?

**Comment by mpustovoi** [comment]
So there is support for shaders 🌄
However, I think that this mod should not be present in the modpack, since it's fully revealed only in a singleplayer
In multiplayer, you will just see the absence of fog near you, since it will take an extremely long time to travel and load chunks

**Comment by TheBossMagnus** [comment]
> So there is support for shaders 🌄 However, I think that this mod should not be present in the modpack, since it's fully revealed only in a singleplayer In multiplayer, you will just see the absence of fog near you, since it will take an extremely long time to travel and load chunks

I don't think this is a big problem, how optifine handle these situations?

**Comment by mpustovoi** [comment]
> I don't think this is a big problem, how optifine handle these situations?

I haven't used optifine since 1.19 came out, so I can't say for sure
I don't think the situations of singleplayer and multiplayer are different in optifine

Without shaders it was the same and I guess it is the same now
With shaders in optifine back then, LODs were rendered not as too bright/dark as Iris did
However, it was impossible to play at 15 fps with medium shaders and Distant Horizons settings

**Comment by surfaceflinger** [comment]
Regarding shaders support, I don't think it should be a concern since Distant Horizons will just show a warning that X shader pack doesn't support DH and turn itself off until shaders are disabled or a compatible pack is picked instead.

Looks like there's also multiverse support now based on similarities but not sure how it would handle voids on servers like skyblock since these are well, similar, and might indicate the same world.

It also saves and caches LODs on servers so it could actually be Bobby alternative

**Comment by cochcoder** [comment]
I played around with this mod in FO and I don't notice any buggy behavior with the zoom mod. It still needs some more work in its current state, but it's shaping up to be a very promising mod.

**Comment by mpustovoi** [comment]
Yes, there was a bug when the nodes were cut off when zooming on them. However, it's no longer there.

Now, there is another one. It's more of an inevitability than a real bug.
With a low render distance, all non-solid blocks in vanilla chunks acquire a pre-LOD state or something. I suppose this was implemented so that a hole wouldn't appear there when unloading these chunks, and instead a LOD chunk seamlessly inserted.

**Comment by cochcoder** [comment]
> With a low render distance, all non-solid blocks in vanilla chunks acquire a pre-LOD state or something.

How low of a render distance did you use to notice this? 

Also, one change that should be made in the config if it's still set to the current default, if this mod is added. Is changing the server lod folder from only using the server name, to server name and address.

**Comment by mpustovoi** [comment]
> How low of a render distance did you use to notice this?

About 8 to have a good fps with shaders.

I'll take a screenshot as soon as I have time.

**Comment by mpustovoi** [comment]
You can see this effect here:
![image](https://github.com/Fabulously-Optimized/fabulously-optimized/assets/92390626/ad7137a2-0788-4089-a0d4-839ffcf81b7b)

**Comment by cochcoder** [comment]
I have noticed that issue, mostly when watching real water turn into LODs and vise versa. I'd imagine that this will be improved upon just as they improved the general look of the LODs from previous versions. Even then, I personally don't find it very noticeable as my eyes move towards the mountains and other structures.

**Comment by mpustovoi** [comment]
Yes, I haven't noticed it for a long time either.
These are the implementation features that will probably remain in their current form for a long time.

**Comment by mpustovoi** [comment]
So the Distant Horizons 2.2.0 came out.
Personally, I really noticed an increase in the speed of creating LODs of the world by about 25% on my heavily modified instance.  This applies to any dimension, vanilla ones, from datapacks or mods.
Also, many problems with LOD chunks in the ocean have really disappeared. Before that, I constantly noticed black spots and oddities in the middle of the ocean or where the monuments and other underwater structures are generated.

I have an i5-7600K CPU and a GTX 1060 of 6 GB (yes, I am so old).
With a vanilla view range of 10 chunks and 64 DH chunks with default settings of the Complementary shaders, I have from 40 to 60 fps. When on the plain or in the ocean, FPS reaches 80-90, and when in the forest or on a mountain it can fell to 30.
Without shaders, with the same view range settings, I had 150-200 fps.

Noisium and C2ME truly increase the loading speed of LODs even more, so my instance with Distant Horizons cannot exist without them.

**Comment by mpustovoi** [comment]
However, there is still a problem that some structures may partially or completely fail to load as LODs, leaving only the LOD terrain instead of themselves.
In my case, this applies to all types of ships from the mod `When Dungeons Arise: Seven Seas`, floating villages from `Tidal Towns` and other structures, some of which are located near the water.

The second problem partially follows the first one.
When trying to load such bugged chunks and then leave them, `Distant Horizons` may not change them at all. In general, this applies to any noticeable change in chunks. For example, after the explosion of a large amount of dynamite or a nuclear bomb in the chunks, and then quickly leaving them.
I do not know exactly how long it is necessary to stay near these chunks within a vanilla view range, but I think 5 minutes will definitely be enough. Otherwise, the algorithm for generating nearby chunks of `Distant Horizons` apparently doesn't have time to prioritize the threads to this.

**Comment by cochcoder** [comment]
I just did a new test with DH (version 2.2.1-a) at settings I personally think are more in line with the default settings of FO. With the render distance set to 6 and DH's render distance set to 32 I get reasonable performance. I get around 76-81 FPS with around 32% of the LODs loaded in. I have a low end laptop and I plan on redoing the performance tests soon. The mods used were: Sodium, EntityCulling, Iris, Lithium, More Culling, FerriteCore, ImmediatelyFast, ModernFix, and FPS-Display. 

I also noticed a significant benefit with Noisium! After 6 minutes at the same settings and seed, the world with Noisium enabled generated a significantly more than the one without. I have two pictures below showing the difference:

Without Noisium: 
![NoNisiom](https://github.com/user-attachments/assets/3f79ea45-e1a4-45b8-a8e0-c8df30d3f54a)

With Noisium: 
![WithNoisium](https://github.com/user-attachments/assets/7b4ac050-121f-4b71-b525-4f9ffcd46dec)

**Comment by cochcoder** [comment]
> However, there is still a problem that some structures may partially or completely fail to load as LODs, leaving only the LOD terrain instead of themselves.
> In my case, this applies to all types of ships from the mod When Dungeons Arise: Seven Seas, floating villages from Tidal Towns and other structures, some of which are located near the water.
> 
> The second problem partially follows the first one.
> When trying to load such bugged chunks and then leave them, Distant Horizons may not update them at all. In general, this applies to any noticeable update in chunks. For example, after the explosion of a large amount of dynamite or a nuke in the chunks, and then quickly leaving them.
> I don't know exactly how long it is necessary to stay near these chunks within a vanilla view range, but I think 5 minutes will definitely be enough. Otherwise, the algorithm for updating nearby LODs apparently doesn't have time to prioritize the threads to this.

Do you still have this issue with version 2.2.1-a? There was a bug were feature generation was broken when Lithium was used.

**Comment by mpustovoi** [comment]
> Do you still have this issue with version 2.2.1-a? There was a bug were feature generation was broken when Lithium was used.

I just looked at the changelog.

However, I used Fabric 1.20.4 for my tests, and the list of changelog:
> - Fix Lithium breaking DH world gen for MC 1.20.1 and older

However, I think here's what will help me:
> - Change delayed LOD modified save from 2 seconds -> 500 ms
>  - This should make LODs appear to update slightly faster

I'll test it myself when I have time.

**Comment by mpustovoi** [comment]
Yeah, this update really helped. I no longer saw partially torn structures, including those on the water.

**Comment by MarthinusBosman** [comment]
I'd really love to know how anyone got DH working alongside FO. Every config I've tried results in broken mod requirements that when dealt with result in the game crashing on startup.

**Comment by MarthinusBosman** [comment]
Actually, just figured out that by disabling MoreCulling it actually runs

**Comment by mpustovoi** [comment]
> Fails because 1.7.4 of Iris is needed, upgrade, needs newer Sodium, upgrade, then remove reese's sodium options and sodium extra's. Then game starts, but crashes whenever I enter a world.

You may also need to delete Indium since it's no longer needed with Sodium 0.6.

> Actually, just figured out that by disabling MoreCulling it actually runs

I don't think that was an actual issue. No clue how did this work.

**Comment by mpustovoi** [comment]
Reese's Sodium Options and Sodium Extra should work then.

> You may also need to delete Indium.

**Comment by MarthinusBosman** [comment]
Yeah, fiddling more, I needed to update MoreCulling instead, but Reese's Sodium Options still breaks it.

**Comment by notrohit1702** [comment]
> I just did a new test with DH (version 2.2.1-a) at settings I personally think are more in line with the default settings of FO.

Could you share your config?

**Comment by felipecrs** [comment]
Maybe wait before Distant Horizons undergoes optimization: https://www.youtube.com/watch?v=MiVtf8_W6JI&t=167s

**Comment by Madis0** [comment]
Based on crash reports users with the mod added have reported to me, there are still issues on different hardware. Therefore, postponing.

**Comment by imveryh4ppy** [comment]
some useless additional information: it can cache chunks

**Comment by TiermoInvin** [comment]
I think it's time to revive this thread since its full release 2 days ago here https://youtu.be/tlqsD3VrMIQ?si=kh7Tv2TfR3LnHDw-
